File: packages/kotlin/src/main/kotlin/ke/locations/KenyaLocations.kt
package ke.locations

import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/**
 * Main entry point for the kenya-locations library.
 *
 * Works on Android, Spring Boot, Ktor, CLI tools, or any Java project —
 * no Android Context or special initialisation required.
 *
 * Data is parsed lazily on first access and cached for the lifetime of the process.
 *
 * Usage:
 * ```kotlin
 * val counties  = KenyaLocations.getCounties()
 * val wards     = KenyaLocations.getWardsInConstituency("Westlands")
 * val results   = KenyaLocations.search("karen")
 * ```
 */
object KenyaLocations {

    private val json = Json { ignoreUnknownKeys = true }

    // ─── Lazy-loaded data (each file parsed only when first accessed) ──────────

    private val _counties: List<County> by lazy {
        loadList("counties.json", ListSerializer(County.serializer()))
    }
    private val _subCounties: List<SubCounty> by lazy {
        loadList("sub-counties.json", ListSerializer(SubCounty.serializer()))
    }
    private val _constituencies: List<Constituency> by lazy {
        loadList("constituencies.json", ListSerializer(Constituency.serializer()))
    }
    private val _wards: List<Ward> by lazy {
        loadList("wards.json", ListSerializer(Ward.serializer()))
    }
    private val _localities: List<Locality> by lazy {
        loadList("locality.json", ListSerializer(Locality.serializer()))
    }
    private val _areas: List<Area> by lazy {
        loadList("area.json", ListSerializer(Area.serializer()))
    }

    // Fast lookup maps — also lazy
    private val countiesByCode: Map<String, County> by lazy { _counties.associateBy { it.code } }
    private val countiesByName: Map<String, County> by lazy { _counties.associateBy { it.name } }
    private val constituenciesByCode: Map<String, Constituency> by lazy { _constituencies.associateBy { it.code } }
    private val constituenciesByName: Map<String, Constituency> by lazy { _constituencies.associateBy { it.name } }

    // ─── Getters ──────────────────────────────────────────────────────────────

    fun getCounties(): List<County> = _counties
    fun getSubCounties(): List<SubCounty> = _subCounties
    fun getConstituencies(): List<Constituency> = _constituencies
    fun getWards(): List<Ward> = _wards
    fun getLocalities(): List<Locality> = _localities
    fun getAreas(): List<Area> = _areas

    fun getCountyByCode(code: String): County? = countiesByCode[code]
    fun getCountyByName(name: String): County? = countiesByName[name]

    fun getConstituencyByCode(code: String): Constituency? = constituenciesByCode[code]
    fun getConstituencyByName(name: String): Constituency? = constituenciesByName[name]

    fun getSubCountiesInCounty(countyName: String): List<SubCounty> =
        _subCounties.filter { it.county == countyName }

    fun getConstituenciesInCounty(countyName: String): List<Constituency> =
        _constituencies.filter { it.county == countyName }

    fun getWardsInConstituency(constituencyName: String): List<Ward> =
        _wards.filter { it.constituency == constituencyName }

    fun getWardsInCounty(countyName: String): List<Ward> {
        val constituencyNames = getConstituenciesInCounty(countyName).map { it.name }.toSet()
        return _wards.filter { it.constituency in constituencyNames }
    }

    fun getLocalitiesInCounty(countyName: String): List<Locality> =
        _localities.filter { it.county == countyName }

    fun getAreasInLocality(localityName: String): List<Area> =
        _areas.filter { it.locality == localityName }

    fun getAreasInCounty(countyName: String): List<Area> =
        _areas.filter { it.county == countyName }

    // ─── Search ───────────────────────────────────────────────────────────────

    /**
     * Fuzzy search across all entity types using a Levenshtein sliding-window algorithm,
     * mirroring the Fuse.js behaviour used on the JS platform (effective threshold ≈ 0.4).
     * Tolerates minor typos — e.g. "Nairob" matches "Nairobi".
     * Results are sorted by relevance score (best match first).
     * Returns up to [limit] results (default 20).
     */
    fun search(query: String, limit: Int = 20): List<SearchResult<*>> {
        val q = query.trim()
        if (q.length < 2) return emptyList()

        val scored = ArrayList<Pair<Double, SearchResult<*>>>()

        fun <T : Any> collect(items: List<T>, type: SearchType, name: (T) -> String) {
            for (item in items) {
                val score = fuzzyScore(q, name(item)) ?: continue
                scored.add(score to SearchResult(type, item))
            }
        }

        collect(_counties, SearchType.COUNTY) { it.name }
        collect(_subCounties, SearchType.SUB_COUNTY) { it.name }
        collect(_constituencies, SearchType.CONSTITUENCY) { it.name }
        collect(_wards, SearchType.WARD) { it.name }
        collect(_localities, SearchType.LOCALITY) { it.name }
        collect(_areas, SearchType.AREA) { it.name }

        return scored.sortedBy { it.first }.take(limit).map { it.second }
    }

    fun searchByType(query: String, type: SearchType, limit: Int = 20): List<SearchResult<*>> =
        search(query, limit * 6).filter { it.type == type }.take(limit)

    // ─── Fuzzy matching ───────────────────────────────────────────────────────

    /**
     * Returns a relevance score in [0.0, 1.0) for [pattern] against [text], or null if
     * no fuzzy match exists. Score 0.0 = exact substring (perfect). Higher = worse match.
     *
     * Algorithm: exact substring check first; then Levenshtein sliding-window with
     * maxErrors = max(1, floor(patternLen × 0.4)). Matches only when the best window
     * score ≤ 0.4, keeping behaviour in line with Fuse.js threshold = 0.4.
     */
    private fun fuzzyScore(pattern: String, text: String): Double? {
        val p = pattern.lowercase()
        val t = text.lowercase()

        if (t.contains(p)) return 0.0
        if (p.length < 2) return null

        val pLen = p.length
        val tLen = t.length
        val maxErrors = maxOf(1, (pLen * 0.4).toInt())
        val windowSize = pLen + maxErrors

        var bestScore = Double.MAX_VALUE

        for (start in 0..maxOf(0, tLen - 1)) {
            val end = minOf(start + windowSize, tLen)
            val dist = levenshteinDistance(p, t.substring(start, end))
            if (dist <= maxErrors) {
                val score = dist.toDouble() / pLen.toDouble()
                if (score < bestScore) bestScore = score
            }
            if (bestScore == 0.0) break
        }

        return if (bestScore <= 0.4) bestScore else null
    }

    /**
     * Space-optimised Levenshtein distance — O(min(m,n)) space.
     */
    private fun levenshteinDistance(s: String, t: String): Int {
        val m = s.length; val n = t.length
        if (m == 0) return n
        if (n == 0) return m
        val dp = IntArray(n + 1) { it }
        for (i in 1..m) {
            var prev = dp[0]; dp[0] = i
            for (j in 1..n) {
                val temp = dp[j]
                dp[j] = if (s[i - 1] == t[j - 1]) prev
                         else 1 + minOf(prev, dp[j], dp[j - 1])
                prev = temp
            }
        }
        return dp[n]
    }

    // ─── Internal ─────────────────────────────────────────────────────────────

    private fun <T> loadList(resource: String, deserializer: kotlinx.serialization.DeserializationStrategy<T>): T {
        val stream = KenyaLocations::class.java.getResourceAsStream("/$resource")
            ?: error("kenya-locations: classpath resource '/$resource' not found. Make sure the library JAR is on the classpath.")
        return stream.bufferedReader().use { json.decodeFromString(deserializer, it.readText()) }
    }
}

File: packages/swift/Sources/KenyaLocations/KenyaLocations.swift
import Foundation

/// Main entry point for the kenya-locations library.
///
/// Works on iOS, macOS, tvOS, watchOS, and any Swift/Apple platform.
/// Data is parsed lazily on first access and cached for the lifetime of the process.
///
/// Usage:
/// ```swift
/// let counties  = KenyaLocations.shared.getCounties()
/// let wards     = KenyaLocations.shared.getWardsInConstituency("Westlands")
/// let results   = KenyaLocations.shared.search("karen")
/// ```
public final class KenyaLocations: @unchecked Sendable {

    public static let shared = KenyaLocations()

    private let decoder = JSONDecoder()
    private let lock = NSLock()

    // Backing storage for lazy-loaded data
    private var _counties: [County]?
    private var _subCounties: [SubCounty]?
    private var _constituencies: [Constituency]?
    private var _wards: [Ward]?
    private var _localities: [Locality]?
    private var _areas: [Area]?

    // Fast lookup maps
    private var _countiesByCode: [String: County]?
    private var _countiesByName: [String: County]?
    private var _constituenciesByCode: [String: Constituency]?
    private var _constituenciesByName: [String: Constituency]?

    private init() {}

    // ─── Getters ──────────────────────────────────────────────────────────────

    public func getCounties() -> [County] { counties }
    public func getSubCounties() -> [SubCounty] { subCounties }
    public func getConstituencies() -> [Constituency] { constituencies }
    public func getWards() -> [Ward] { wards }
    public func getLocalities() -> [Locality] { localities }
    public func getAreas() -> [Area] { areas }

    public func getCountyByCode(_ code: String) -> County? { countiesByCode[code] }
    public func getCountyByName(_ name: String) -> County? { countiesByName[name] }

    public func getConstituencyByCode(_ code: String) -> Constituency? { constituenciesByCode[code] }
    public func getConstituencyByName(_ name: String) -> Constituency? { constituenciesByName[name] }

    public func getSubCountiesInCounty(_ countyName: String) -> [SubCounty] {
        subCounties.filter { $0.county == countyName }
    }

    public func getConstituenciesInCounty(_ countyName: String) -> [Constituency] {
        constituencies.filter { $0.county == countyName }
    }

    public func getWardsInConstituency(_ constituencyName: String) -> [Ward] {
        wards.filter { $0.constituency == constituencyName }
    }

    public func getWardsInCounty(_ countyName: String) -> [Ward] {
        let constituencyNames = Set(getConstituenciesInCounty(countyName).map { $0.name })
        return wards.filter { constituencyNames.contains($0.constituency) }
    }

    public func getLocalitiesInCounty(_ countyName: String) -> [Locality] {
        localities.filter { $0.county == countyName }
    }

    public func getAreasInLocality(_ localityName: String) -> [Area] {
        areas.filter { $0.locality == localityName }
    }

    public func getAreasInCounty(_ countyName: String) -> [Area] {
        areas.filter { $0.county == countyName }
    }

    // ─── Search ───────────────────────────────────────────────────────────────

    /// Case-insensitive substring search across all entity types.
    /// Returns up to `limit` results (default 20).
    ///
    /// - Note: This currently uses substring matching. To align with the JS
    ///   platform (Fuse.js Bitap, threshold 0.2), replace this body with the
    ///   `Ifrit` package (https://github.com/ukushu/Ifrit) using the same
    ///   threshold and `keys: ["name"]` configuration.
    ///   See TODO(search-parity) in KenyaLocations.kt for the Kotlin equivalent.
    public func search(_ query: String, limit: Int = 20) -> [SearchResult] {
        let q = query.lowercased()
        var results: [SearchResult] = []
        counties.filter { $0.name.lowercased().contains(q) }.forEach { results.append(.county($0)) }
        subCounties.filter { $0.name.lowercased().contains(q) }.forEach { results.append(.subCounty($0)) }
        constituencies.filter { $0.name.lowercased().contains(q) }.forEach { results.append(.constituency($0)) }
        wards.filter { $0.name.lowercased().contains(q) }.forEach { results.append(.ward($0)) }
        localities.filter { $0.name.lowercased().contains(q) }.forEach { results.append(.locality($0)) }
        areas.filter { $0.name.lowercased().contains(q) }.forEach { results.append(.area($0)) }
        return Array(results.prefix(limit))
    }

    public func searchByType(_ query: String, type: SearchType, limit: Int = 20) -> [SearchResult] {
        search(query, limit: limit * 6).filter { $0.type == type }.prefix(limit).map { $0 }
    }

    // ─── Thread-safe lazy accessors ───────────────────────────────────────────

    private var counties: [County] {
        lock.lock(); defer { lock.unlock() }
        if let v = _counties { return v }
        let v: [County] = load("counties")
        _counties = v; return v
    }

    private var subCounties: [SubCounty] {
        lock.lock(); defer { lock.unlock() }
        if let v = _subCounties { return v }
        let v: [SubCounty] = load("sub-counties")
        _subCounties = v; return v
    }

    private var constituencies: [Constituency] {
        lock.lock(); defer { lock.unlock() }
        if let v = _constituencies { return v }
        let v: [Constituency] = load("constituencies")
        _constituencies = v; return v
    }

    private var wards: [Ward] {
        lock.lock(); defer { lock.unlock() }
        if let v = _wards { return v }
        let v: [Ward] = load("wards")
        _wards = v; return v
    }

    private var localities: [Locality] {
        lock.lock(); defer { lock.unlock() }
        if let v = _localities { return v }
        let v: [Locality] = load("locality")
        _localities = v; return v
    }

    private var areas: [Area] {
        lock.lock(); defer { lock.unlock() }
        if let v = _areas { return v }
        let v: [Area] = load("area")
        _areas = v; return v
    }

    private var countiesByCode: [String: County] {
        lock.lock(); defer { lock.unlock() }
        if let v = _countiesByCode { return v }
        let v = Dictionary(uniqueKeysWithValues: (_counties ?? []).map { ($0.code, $0) })
        _countiesByCode = v; return v
    }

    private var countiesByName: [String: County] {
        lock.lock(); defer { lock.unlock() }
        if let v = _countiesByName { return v }
        let v = Dictionary(uniqueKeysWithValues: (_counties ?? []).map { ($0.name, $0) })
        _countiesByName = v; return v
    }

    private var constituenciesByCode: [String: Constituency] {
        lock.lock(); defer { lock.unlock() }
        if let v = _constituenciesByCode { return v }
        let v = Dictionary(uniqueKeysWithValues: (_constituencies ?? []).map { ($0.code, $0) })
        _constituenciesByCode = v; return v
    }

    private var constituenciesByName: [String: Constituency] {
        lock.lock(); defer { lock.unlock() }
        if let v = _constituenciesByName { return v }
        let v = Dictionary(uniqueKeysWithValues: (_constituencies ?? []).map { ($0.name, $0) })
        _constituenciesByName = v; return v
    }

    // ─── Internal ─────────────────────────────────────────────────────────────

    private func load<T: Decodable>(_ resource: String) -> [T] {
        guard let url = Bundle.module.url(forResource: resource, withExtension: "json") else {
            fatalError("kenya-locations: resource '\(resource).json' not found in bundle. Make sure to run the copy-data script before building.")
        }
        do {
            let data = try Data(contentsOf: url)
            return try decoder.decode([T].self, from: data)
        } catch {
            fatalError("kenya-locations: failed to decode '\(resource).json': \(error)")
        }
    }
}
